readme.md
# Gateway Demo Setup
## guide
<https://software-dl.ti.com/jacinto7/esd/processor-sdk-rtos-j7200/07_03_00_07/exports/docs/gateway-demos/docs/Can-Eth-User_Guide.html>

<https://software-dl.ti.com/jacinto7/esd/processor-sdk-rtos-jacinto7/08_00_00_12/exports/docs/gateway-demos/docs/Can-Eth-User_Guide.html>

## build the demo 
1. download the startJacinto tools. 
2. run `source env*` 
3. install the PSDKRA. 
4. run `gateway-build`
5. image located in the below 
>  sdk/gateway-demos/binary/can_eth_gateway_app/bin/j721e_evm/can_eth_gateway_app_mcu2_1_release.xer5f
>  sdk/ethfw/out/J721E/R5F/SYSBIOS/release/app_remoteswitchcfg_server.xer5f

## pcan tools install
1. run `make ubuntu-install-pcan-tools`
2. run `make gateway-install-pc-tools`
3. modify the configuration file.(~/startJacinto/tools/peak-linux-driver-8.9.3/test/pctools/app_config.txt)
```
#This is a sample App config. You can use this as a template to create your own configuration.

#The PCAN connection below is for MCAN4. Run lspcan command with the PCAN-USB tool connected, to find out the name of the connection
/dev/pcanusbfd32 - 0xD0
#The PCAN connection below is for MCAN9
/dev/pcanusbfd33 - 0xE0
#This is the primary Ethernet interface. It can be anything but recommendation is to use the built-in Ethernet adapter
enxd037455f8444 - d0:37:45:5f:84:44
#This is the secondary Ethernet interface.
enp1s2 - 00:ad:24:8f:f6:4e
#This is the MAC address for Gateway. You can get this from the CCS or UART logs.
gateway - 70:ff:76:1d:88:32

# Remove all comments starting with # before using this file
```
## Run the demo
### CCS Boot Eemo
1. CCS load server and canapp image to MAINR5F_0 and MAINR5F_1, MAINR5F_0 run first.

#### Run CAN2ETH demo
```
peak-linux-driver-8.9.3/test$ ./pcanfdtst tx ETH_ONLY  --bitrate 1000000 --dbitrate 5000000 --clock 80000000 -n 1 -ie 0xD0 --fd -l 64 --tx-pause-us 1000 /dev/pcanusbfd32
peak-linux-driver-8.9.3/test/pctools$ sudo ./recv_1722.out --eth_interface eno1 --timeout 12 --pipes 0 --verbosity verbose
```
Run the demo that based startJacinto
```
make ...
```
#### Run ETH2ETH demo
```
peak-linux-driver-8.9.3/test/pctools$ sudo ./send_1722.out --eth_interface eno1 --gateway_mac 70:ff:76:1d:9f:4f --dst_mac ec:f4:bb:15:36:a4 --ipg 1000 --route ETH --num_packets 10
peak-linux-driver-8.9.3/test/pctools$ sudo ./recv_1722.out --eth_interface eno1 --timeout 12 --pipes 0 --verbosity verbose
```
Run the demo that based startJacinto
```
make ...
```

# LOG 
## MCU_log  r5f_0

```
=======================================================
           CPSW Ethernet Firmware Demo             
=======================================================
ETHFW Version: 0. 1. 1
ETHFW Build Date (YYYY/MMM/DD):2020/Feb/17
ETHFW Commit SHA:ETHFW PermissionFlag:0x1ffffff, UART Connected:true,UART Id:2IPC_echo_test (core : mcu2_0) .....
CPSW_9G Test on MAIN NAVSS
Remote demo device (core : mcu2_0) .....
CpswPhy_bindDriver: PHY 12: OUI:080028 Model:23 Ver:01 <-> 'dp83867' : OK
CpswPhy_bindDriver: PHY 0: OUI:080028 Model:23 Ver:01 <-> 'dp83867' : OK
CpswPhy_bindDriver: PHY 3: OUI:080028 Model:23 Ver:01 <-> 'dp83867' : OK
CpswPhy_bindDriver: PHY 15: OUI:080028 Model:23 Ver:01 <-> 'dp83867' : OK

PHY 0 is alive
PHY 3 is alive
PHY 12 is alive
PHY 15 is alive
PHY 16 is alive
PHY 17 is alive
PHY 18 is alive
PHY 19 is alive
PHY 23 is alive
Host MAC address: 70:ff:76:1d:9f:4e
[NIMU_NDK] CPSW has been started successfully
Function:CpswProxyServer_attachExtHandlerCb,HostId:4,CpswType:1
Function:CpswProxyServer_ioctlHandlerCb,HostId:4,Handle:a2cee3f4,CoreKey:38acb976, Cmd:20000,InArgsLen:24, OutArgsLen:4 
Function:CpswProxyServer_registerMacHandlerCb,HostId:4,Handle:a2cee3f4,CoreKey:38acb976, MacAddress:70:ff:76:1d:9f:4f, FlowIdx:178, FlowIdxOffset:6
Cpsw_ioctlInternal: CPSW: Registered MAC address.ALE entry:10, Policer Entry:0Cpsw_handleLinkUp: port 3: Link up: 1-Gbps Full-Duplex
Cpsw_handleLinkUp: port 2: Link up: 1-Gbps Full-Duplex
Function:CpswProxyServer_clientNotifyHandlerCb,HostId:4,Handle:38acb976,CoreKey:a2cee3f4,NotifyId:RPMSG_KDRV_TP_ETHSWITCH_CLIENTNOTIFY_CUSTOM,NotifyLen
```
## MCU_log  r5f_1
```
[MAIN_Cortex_R5_0_1] 
Gateway App:Variant - Pre Compile being used !!!
CAN_APP: Successfully Enabled CAN Transceiver Main Domain Inst 4,9,11!!!
 
Gateway App: CDD IPC MCAL Version Info
Gateway App:---------------------
Gateway App: Vendor ID           : 44
Gateway App: Module ID           : 255
Gateway App: Starting CAN Rx
Gateway App: Starting CAN Rx Interface
Gateway App: SW Major Version    : 1
Gateway App: SW Minor Version    : 0
Gateway App: SW Patch Version    : 0
 
Gateway App:MAC Port 1 Address: 70:ff:76:1d:9f:4f
Gateway App: Starting Eth Rx Interface
Gateway App: Starting Gateway Router
```




driver/release_notes.txt
2011-01-12 - 6.24 - content:
- handle PCMCIA cards in 2.6.35 kernel

2011-01-18 - 6.24.1 - content:
- support 2.6.36 kernel

2011-01-18 - 6.24.2 - content:
- support 2.6.37 kernel

2011-03-11 - 7.0 - content:
- support PCAN-USB-PRO 
- fix 2.6.37 with PCMCIA
- handle timestamps from PCAN-USB too

2011-03-17 - 7.1 - content:
- support of dynamic usb minors allocation (kernel config option)

2011-03-24 - 7.1.1 - content:
- fix typo in src/pcan_dongle.c to be able to use the PCAN-Dongle

2011-04-08 - 7.1.2 - content:
- fix sja1000 mutual exclusion in write() functions

2011-04-21 - 7.2 - content:
- support 2.6.38.3 kernel

2011-07-19 - 7.2.1 - content:
- support of older kernel which don't define list_for_each_prev_safe and
  usb_driver structure don't define 'drvwrap' field (2.6.18 for example).
- fix - non - support of pc-card even if PCCARD_SUPPORT is defined on command
  line.

2011-07-26 - 7.2.2 - content:
- fix pcan usb regression issue (device number now back in irq field/column)
- add support for kernel 3.0

2011-07-27 - 7.3 - content:
- include all 7.2.x patches
- add support for OpenSUSE 11.3 x86_64

2011-07-28 - 7.3 - content:
- new delivery because of OpenSUSE config issue (missing i2c and pcmcia modules)
  while options are set in the kernel! So this delivery is quite the same than
  2011-07-27 but enforces some kernel config check and removes a patch of a
  customer that brought some warning on 64-bits compilation.
- the 2011-07-27 delivery should not be used and will be renamed in svn repo
  into "Release_20110727_n-7.3-dontuse"

2011-09-12 - 7.4 - content:
- remove any references to selecttest into test/Makefile
- fix USB-PRO on odd number of frames sent in between insmod/rmmod cycle
- add cancelation of calibration message on device removing 
- move can off messages sending on channel closing rather than on device 
  removing

2011-10-05 - 7.4.1 - content:
- fix usb-pro packet size issue against receiving messages with size greater 
  than 512 bytes

2011-10-06 - 7.4.2 - content:
- handle specifically USB-PRO error status and record fragmentation w/ packet
  size < 1024 bytes
- add error messages in pci probing to ease debugging
- handle big/little endian in USB-PRO message coding/decoding

2012-01-18 - 7.5 - content:
- handle new PCAN-miniPCI-Express
- fix potential issue in PCIeC driver by using a delayed work instead of a timer
  to control the leds (i2c_transfer() may sleep because of underlying drivers).
- fix some crashes when PCIeC support was requested
- fix driver build until v3.03
- fix potential infinite loop in usb driver part
- fix wrong /dev/NULL ref in driver Makefile 
- make install now adds peak_xxx mainline drivers to system blacklist (see 
  udev/blacklist-peak.conf) and forces their unloading from memory
- fix potential issue with PCAN-USB fw 2.8
- add "uninstall" target to Makefiles to properly remove driver from the system

2012-05-22 - 7.6 - content:
- add support for some new PCI devices
- remove unused MPCIE_SUPPORT compilation option from Makefile
- fix compilation issue with old kernel versions and the usage of 
  net_ratelimit()

2012-07-26 - 7.7 - content:
- fix compilation issues with Kernel versions until 3.5 (Thx Oliver H. ;-)
- PCAN-USB Pro: fix compilation issue with undefined symbol "pr_err()" and
(very) old Kernels (2.6.22).

2013-01-31 - 7.8 - content:
- handle "version.h" moving in Kernel 3.7+
- pcan_netdev.c: add support for Kernel 3.9+ linux-can
- add DESTDIR support in driver/lib/test Makefile
- handle peak_pci removing according to Kernel version (< 3.4 crash when rmmod)
- add build date/time in /proc/pcan display
- dev_register(pcan_main_linux.c): add forgotten HW_ISA_SJA type when creating
  and removing nodes with UDEV_SUPPORT
- pcan_usb_setCANOff(pcan_usb.c): always force usb_clear_halt() and mdelay(50)
  after CAN=off: this fixes an issue writing a single msg then closing the 
  device next.
- transmittest.cpp: fix wrong errno usage calling CAN_Write
- receivetest.c: fix wrong errno usage calling LINUX_CAN_Read

2013-04-04 - 7.8.1 - content:
- src/sja1000.c: remove useless test on RECEIVE_MSG_COUNTER

2013-04-15 - 7.8.2 - content:
- add mutex to control access to devices list; add two functions to safe 
  add/del items to/from this list and rescan all sources to use these functions,
  or at least, to access devices list with respect to  mutual exclusion.
- src/sja1000.c: reverted to 7.8
- src/pcan_main.c: rework remove_dev_list() to remove device EVEN if no
  cleanup() callback is regigistered (potential memory cosumption)

2013-04-16 - 7.8.3 - content:
- src/pcan_main.c: consolidate pcan_read_procmem() against NULL "netdev" pointer
  access, when NETDEV_SUPPORT is defined
- add mutex lock/unlock around all other device list accesses (opportunity to 
  remove dust in the C coding style of some files)

2013-07-03 - 7.8.4 - content:
- handle /proc/pcan read access in a more modern way (create_proc_entry() 
  deprecated since 3.9)

2013-08-07 - 7.9-beta
- src/pcan_usb.c: handle the fact that DLC field in PCAN-USB protocol can 
  be > 8. For the moment, 'dlc' fields in user C structs is not!

2013-08-13 - 7.9-local
- fix bug in first column of /proc/pcan when displaying USB device can channel
- also moved to more efficient way of filling and returning struct objects in
  all "pcan_ioctl_diag_xxx()" functions

2013-08-14 - 7.9
- official release tested with 3.8 and 3.10 kernels

2013-10-07 - 7.9.1-beta
- added new PCI ID 0x0009 (PCAN-PCI Express OEM)

2013-10-10 - 7.9.2-beta
- handle USB error like BUSOFF to force application to restart

2013-10-16 - 7.10.0-beta
- handle usb core error in read callback (interrupt context) by properly 
  closing the device instead of ignoring the error
- handle USB error like BUSOFF to force application to restart
- remove "PEAK-Systems" ending "s" to MODULE_DESCRIPTION string 

2014-01-21 - 7.10 - content
- official release tested with 3.10 kernel

2014-04-30 - 7.11 - content

driver/src/pcan_common.h
/* SPDX-License-Identifier: GPL-2.0 */
/*****************************************************************************
 * Copyright (C) 2001-2010  PEAK System-Technik GmbH
 *
 * linux@peak-system.com
 * www.peak-system.com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * Maintainer(s): Stephane Grosjean (s.grosjean@peak-system.com)
 *
 * Major contributions by:
 *                Edouard Tisserant (edouard.tisserant@lolitech.fr) XENOMAI
 *                Laurent Bessard   (laurent.bessard@lolitech.fr)   XENOMAI
 *                Oliver Hartkopp   (oliver.hartkopp@volkswagen.de) socketCAN
 *                Klaus Hitschler   (klaus.hitschler@gmx.de)
 *
 *****************************************************************************/

/*****************************************************************************
 *
 * global defines to include in all files this module is made of
 * ! this must always the 1st include in all files !
 *
 * $Id$
 *
 *****************************************************************************/

#ifndef __PCAN_COMMON_H__
#define __PCAN_COMMON_H__

#ifndef __KERNEL__
#define __KERNEL__
#endif
#ifndef MODULE
#define MODULE
#endif

#include <linux/kernel.h>
#include <linux/slab.h>

/* if this file is not found: please look @ /boot/vmlinuz.version.h and make a
 * symlink */
#include <linux/version.h>

#include <linux/module.h>
#include <linux/stringify.h>

/* support for MODVERSIONS */
#ifndef AUTOCONF_INCLUDED
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,38)
#include <generated/autoconf.h>
#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,16)
#include <linux/autoconf.h>
#else
#include <linux/config.h>
#endif
#define AUTOCONF_INCLUDED
#endif

#if defined(CONFIG_MODVERSIONS) && !defined(MODVERSIONS)
#define MODVERSIONS
#endif

#ifdef MODVERSIONS
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,0)
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,10)
#include <config/modversions.h>
#endif
#else
#include <linux/modversions.h>
#endif
#endif

//#define DEBUG_MEM

/* to get smoothly into udev */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,17)
#ifdef NO_RT
#define UDEV_SUPPORT
#else
#undef UDEV_SUPPORT
#endif
#endif

#if LINUX_VERSION_CODE < KERNEL_VERSION(4,15,0)
#define pcan_setup_timer(a, b, c)	setup_timer(a, b, c)
#else
#define pcan_setup_timer(a, b, c)	timer_setup(a, b, 0)
#endif

/* #if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,35) */
#ifndef pr_warn
#define pr_warn(fmt, ...)	printk(KERN_WARNING pr_fmt(fmt), ##__VA_ARGS__)
#endif
#ifndef pr_err
#define pr_err(fmt, ...)	printk(KERN_ERR pr_fmt(fmt), ##__VA_ARGS__)
#endif

#ifndef NO_COMPAT
#ifdef CONFIG_COMPAT
#define PCAN_CONFIG_COMPAT
#endif
#endif

/* if defined, clock drift between CPU clock and CAN clock is handled.
 * if not defined, HWTIMESTAMP is made of a host time base and an offset 
 * computed from the device clock ticks. Note that such timestamps might give
 * time from the future. */
#define PCAN_HANDLE_CLOCK_DRIFT	

/* helper to compare version numbers */
#define VER_NUM(x, y, z)	(((u32 )(x) << 24) | \
				 ((u32 )(y) << 16) | \
				 ((u32 )(z) << 8))
#define VER_MAJ(x)		(((x) >> 24) & 0xff)
#define VER_MIN(x)		(((x) >> 16) & 0xff)
#define VER_SUB(x)		(((x) >> 8)  & 0xff)

static inline u64 timeval_to_us(struct timeval *tv)
{
	return ((u64 )tv->tv_sec * USEC_PER_SEC) + tv->tv_usec;
}

static inline void timeval_add_us(struct timeval *tv, signed long us)
{
	tv->tv_usec += us;

	while (tv->tv_usec < 0) {
		tv->tv_usec += USEC_PER_SEC;
		tv->tv_sec--;
	}

	tv->tv_sec += tv->tv_usec / USEC_PER_SEC;
	tv->tv_usec %= USEC_PER_SEC;
}

static inline signed long timeval_diff(struct timeval *tv0, struct timeval *tv1)
{
	return (long )(timeval_to_us(tv0) - timeval_to_us(tv1));
}

static inline int timeval_cmp(struct timeval *tv0, struct timeval *tv1)
{
	return (tv0->tv_sec != tv1->tv_sec) ? tv0->tv_sec - tv1->tv_sec
						: tv0->tv_usec - tv1->tv_usec;
}

#define timeval_is_older(t1, t2)		timeval_cmp(t1, t2) < 0

#ifdef NO_RT

#if LINUX_VERSION_CODE < KERNEL_VERSION(3,17,0)
#include <linux/hrtimer.h>
#endif

#define PCAN_IRQF_SHARED		IRQF_SHARED

static inline int pcan_task_can_wait(void)
{
	return 1;
}

static inline u64 pcan_getnow_ns(void)
{
#if LINUX_VERSION_CODE < KERNEL_VERSION(3,17,0)
	return ktime_to_ns(ktime_get());
#else
	return ktime_get_ns();
#endif
}

static inline void pcan_getnow_ts(struct timespec *ts, u64 *ptv_ns)
{
	getrawmonotonic(ts);
	if (ptv_ns)
		*ptv_ns = pcan_getnow_ns();
}

#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,0,0)
static inline void do_gettimeofday(struct timeval *tv)
{
	struct timespec64 now;

	ktime_get_real_ts64(&now);
	tv->tv_sec = now.tv_sec;
	tv->tv_usec = now.tv_nsec/1000;
}
#endif

static inline void pcan_gettimeofday_ex(struct timeval *tv, u64 *ptv_ns)
{
#ifdef USES_MONOTONIC_CLOCK
	struct timespec ts;

	pcan_getnow_ts(&ts, ptv_ns);

	//do_div(ts64.tv_nsec, NSEC_PER_USEC);
	tv->tv_usec = ts.tv_nsec / NSEC_PER_USEC;
	tv->tv_sec = ts.tv_sec;
#else
	do_gettimeofday(tv);
	if (ptv_ns)
		*ptv_ns = pcan_getnow_ns();
#endif
}

#include <linux/uaccess.h>

static inline int pcan_copy_from_user(void *to, const void __user *from,
					int size, void *c)
{
	return copy_from_user(to, from, size);
}

static inline int pcan_copy_to_user(void __user *to, const void *from,
					size_t size, void *c)
{
	return copy_to_user(to, from, size);
}

#else

#ifdef XENOMAI3
#include <rtdm/driver.h>

typedef struct rtdm_fd rtdm_user_info_t;
#else
#include <rtdm/rtdm_driver.h>
#endif
#include <asm/div64.h>

#define PCAN_IRQF_SHARED		RTDM_IRQTYPE_SHARED

static inline int pcan_task_can_wait(void)
{
	return rtdm_in_rt_context();
}

#if RTDM_API_VER < 5
typedef nanosecs_abs_t uint64_t;
#endif

static inline u64 pcan_getnow_ns(void)
{
	return rtdm_clock_read();
}

static inline void pcan_gettimeofday_ex(struct timeval *tv, u64 *ptv_ns)
{
	nanosecs_abs_t current_time = rtdm_clock_read();

	if (ptv_ns)
		*ptv_ns = current_time;

	/* a = do_div(b, c); <=> { a = b % c ; b = b / c; } */
	tv->tv_usec = (do_div(current_time, NSEC_PER_SEC) / NSEC_PER_USEC);
	tv->tv_sec = current_time;
}

static inline int pcan_copy_from_user(void *to, const void __user *from,
					int size, void *c)
{
	if (c) {
		if (!rtdm_read_user_ok((rtdm_user_info_t *)c, from, size) ||
		    rtdm_copy_from_user((rtdm_user_info_t *)c, to, from, size))
			return -EFAULT;

		return 0;
	}

	return copy_from_user(to, from, size);
}

static inline int pcan_copy_to_user(void __user *to, const void *from,
					size_t size, void *c)
{
	if (c) {
		if (!rtdm_rw_user_ok((rtdm_user_info_t *)c, to, size) ||
		    rtdm_copy_to_user((rtdm_user_info_t *)c, to, from, size))
			return -EFAULT;

		return 0;
	}

	return copy_to_user(to, from, size);
}
#endif

static inline void pcan_gettimeofday(struct timeval *tv)
{
	pcan_gettimeofday_ex(tv, NULL);
}

/* support for PARPORT_SUBSYSTEM */
#if !defined(CONFIG_PARPORT_MODULE) && !defined(CONFIG_PARPORT) && defined(PARPORT_SUBSYSTEM)
#undef PARPORT_SUBSYSTEM
#endif

/* support for USB */
#if !defined(CONFIG_USB_MODULE) && !defined(CONFIG_USB) && defined(USB_SUPPORT)
#undef USB_SUPPORT
#endif

/* support for PCCARD */
#if !defined(CONFIG_PCMCIA_MODULE) && !defined(CONFIG_PCMCIA) && !defined(CONFIG_PCCARD) && defined(PCCARD_SUPPORT)
#undef PCCARD_SUPPORT
#endif

/* support for PCIe (need I2C algo) */
#if !defined(CONFIG_I2C_ALGOBIT) && !defined(CONFIG_I2C_ALGOBIT_MODULE)
#undef PCIEC_SUPPORT
#endif

/* hotplug may need mutex locking around the global devices list.
 * On the other hand, don't add useless mechanism in the other case. */
#if defined(USB_SUPPORT) || defined(PCCARD_SUPPORT) || defined(NETDEV_SUPPORT) || defined(UDEV_SUPPORT)
#define HANDLE_HOTPLUG
#endif

/* support only versions 2.4.x and 2.6.x */
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,4,0)
#error "This kernel is too old and not supported"
#else
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,0)
#define LINUX_24 /* >= LINUX 2.4.x && < LINUX 2.6 */
#else
#define LINUX_26 /* >= LINUX 2.6 */
#endif
#endif

/* moved from pcan_main.h */
#define DEVICE_NAME	"pcan"	/* the name of the module and the proc entry */

#define kHz				1000
#define MHz				(1000*kHz)
#define GHz				(1000*MHz)

#ifdef DEBUG_MEM
static inline void *_pcan_malloc(char *f, int l, size_t size, gfp_t flags)
{
	void *p = kmalloc(size, flags);
	pr_info(DEVICE_NAME ": %s(%u): alloc(%u): %p\n",
		f, l, (unsigned )size, p);
	return p;
}
#define pcan_malloc(a, b)		_pcan_malloc(__FILE__, __LINE__, a, b)

static inline void *_pcan_free(char *f, int l, const void *p)
{
	pr_info(DEVICE_NAME ": %s(%u): free(%p)\n",
		__FILE__, __LINE__, p);
	kfree(p);
	return NULL;
}

#define pcan_free(p)			_pcan_free(__FILE__, __LINE__, p);

#else
#define pcan_malloc(a, b)		kmalloc(a, b)
static inline void *pcan_free(void *p)
{
	kfree(p);
	return NULL;
}
#endif /* DEBUG_MEM */

/* some preparative definitions for kernel 2.6.x */
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,4,23)
typedef void irqreturn_t;
#define IRQ_NONE
#define IRQ_HANDLED
#define IRQ_RETVAL(x)
#endif

/* IRQ's return val for Linux and RealTime */
#ifdef NO_RT
#define PCAN_IRQ_NONE			IRQ_NONE
#define PCAN_IRQ_HANDLED		IRQ_HANDLED
#define PCAN_IRQ_RETVAL(x)		IRQ_RETVAL(x)
#else
#define PCAN_IRQ_NONE			RTDM_IRQ_NONE
#define PCAN_IRQ_HANDLED		RTDM_IRQ_HANDLED
#define PCAN_IRQ_RETVAL(x)		x
#endif

/* spinlock and mutex common interface */
#ifdef NO_RT
typedef unsigned long	pcan_lock_irqsave_ctxt;
typedef spinlock_t	pcan_lock_t;

#define pcan_lock_init(l)		spin_lock_init(l)
#define pcan_lock_get_irqsave(l, f)	spin_lock_irqsave(l, f)
#define pcan_lock_put_irqrestore(l, f)	spin_unlock_irqrestore(l, f)
#define pcan_lock_try_irqsave(l, f)	spin_trylock_irqsave(l, f)
#define pcan_lock_destroy(l)

typedef struct mutex	pcan_mutex_t;

#define pcan_mutex_init(m)		mutex_init(m)
#define pcan_mutex_lock(m)		mutex_lock(m)
#define pcan_mutex_unlock(m)		mutex_unlock(m)
#define pcan_mutex_trylock(m)		mutex_trylock(m)
#define pcan_mutex_destroy(m)		mutex_destroy(m)

#else
typedef rtdm_lockctx_t	pcan_lock_irqsave_ctxt;
typedef rtdm_lock_t	pcan_lock_t;

#define pcan_lock_init(l)		rtdm_lock_init(l)
#define pcan_lock_get_irqsave(l, f)	rtdm_lock_get_irqsave(l, f)
#define pcan_lock_put_irqrestore(l, f)	rtdm_lock_put_irqrestore(l, f)
#define pcan_lock_destroy(l)

typedef rtdm_mutex_t	pcan_mutex_t;

#define pcan_mutex_init(m)		rtdm_mutex_init(m)
#define pcan_mutex_lock(m)		rtdm_mutex_lock(m)
#define pcan_mutex_unlock(m)		rtdm_mutex_unlock(m)
#define pcan_mutex_trylock(m)		(!rtdm_mutex_timedlock(m, -1, NULL))
#define pcan_mutex_destroy(m)		rtdm_mutex_destroy(m)

#endif

/* event/queue common interface */
#ifdef NO_RT
typedef wait_queue_head_t	pcan_event_t;

#define pcan_event_init(e, v)	init_waitqueue_head(e)
#define pcan_event_free(e)
#define pcan_event_signal(e)	wake_up_interruptible(e)
#define pcan_event_wait_timeout(e, c, t)				\
({									\
	int err;							\
									\
	if (t) {							\
		err = wait_event_interruptible_timeout(e, c,		\
						msecs_to_jiffies(t));	\
		if (!err)						\
			err = -ETIMEDOUT;				\
	} else {							\
		err = wait_event_interruptible(e, c);			\
	}								\
	err;								\
})

#define pcan_event_wait(e, c)	wait_event_interruptible(e, c)
#else
typedef rtdm_event_t		pcan_event_t;

#define pcan_event_init(e, v)	rtdm_event_init(e, v)
#define pcan_event_free(e)	rtdm_event_destroy(e)
#define pcan_event_signal(e)	rtdm_event_signal(e)
#define pcan_event_clear(e)	rtdm_event_clear(e)
#define pcan_event_wait_timeout(e, c, t)				\
({									\
	nanosecs_rel_t ns = (nanosecs_rel_t )t * 1000000LL;		\
	rtdm_toseq_t ts, *pts = &ts;					\
	int err = 0;							\
									\
	if (t)								\
		rtdm_toseq_init(&ts, ns);				\
	else								\
		pts = NULL;						\
	while (!(c)) {							\
		err = rtdm_event_timedwait(&e, ns, pts);		\
		if (err < 0)						\
			break;						\
	}								\
	err;								\
})
#define pcan_event_wait(e, c)	pcan_event_wait_timeout(e, c, 0)
#endif

/* count of function variables changed from 2.6.19 */
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,19)
#define IRQHANDLER(name, p1, p2, p3)	name(p1, p2, p3)
#else
#define IRQHANDLER(name, p1, p2, p3)	name(p1, p2)
#endif

/* switch to disable all printks for not debugging */
#ifdef DEBUG
#define DPRINTK				printk
#else
#define DPRINTK(stuff...)
#endif

/* to manage differences between kernel versions */
int ___request_region(unsigned long from, unsigned long length,
		const char *name);

/* follow current interrupt definition changes */
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,18)
#define IRQF_DISABLED			SA_INTERRUPT
#define IRQF_SHARED			SA_SHIRQ
#endif

/* different data sink alternatives */
#ifdef NETDEV_SUPPORT
#define pcan_xxxdev_rx(d, f)		pcan_netdev_rx(d, f)
#else

/* if defined, STATUS[PCANFD_BUS_LOAD] message is periodically pushed by the
 * driver instead of by the device. This enables to control flooding of
 * the rx queue. */
#define PCAN_USE_BUS_LOAD_TIMER

#define pcan_xxxdev_rx(d, f)		pcan_chardev_rx(d, f)
#endif

#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24)
/* This has been added in 2.6.24 */
#define list_for_each_prev_safe(pos, n, head) \
			for (pos = (head)->prev, n = pos->prev; \
				prefetch(pos->prev), pos != (head); \
						pos = n, n = pos->prev)
#endif

/* set here the current release of the driver 'Release_date_nr' synchronous
 * with SVN
 */
#define CURRENT_RELEASE                 "Release_20191029_n"
#define PCAN_VERSION_MAJOR              8
#define PCAN_VERSION_MINOR              9
#define PCAN_VERSION_SUBMINOR           3
#define CURRENT_VERSIONSTRING		__stringify(PCAN_VERSION_MAJOR)"."\
					__stringify(PCAN_VERSION_MINOR)"."\
					__stringify(PCAN_VERSION_SUBMINOR)

#endif
